In [1]:
from __future__ import print_function, division

In [2]:
from collections import defaultdict
import csv
from datetime import datetime
import matplotlib.pyplot as plt
import pandas as pd
import os

%matplotlib inline

'path' is where all the data files are on my computer. Another way to get the data without downloading them manually is below.

In [3]:
path = '/Users/aleksandra/ds/metis/Project-Benson/data_mta/'

In [4]:
def read_data(path):
    '''opens all txt files in the given directory and creates a df'''
    all_data = pd.DataFrame()
    for filename in os.listdir(path):
        with open(path+str(filename)) as f:
            df = pd.read_csv(f)
            all_data = all_data.append(df)
    all_data.columns = ['C/A', 'UNIT', 'SCP', 'STATION', 'LINENAME', 'DIVISION', 'DATE', 'TIME','DESC', 'ENTRIES', 'EXITS']
    return all_data

def save_as_csv(df, path):
    '''saves df in given directory'''
    df.to_csv(path+'all_data.csv')

Below is the Metis code to get all the data and put them in a data frame. 

In [5]:
def get_data(week_nums):
    '''NOT MY CODE. downlaods all files in the list of weeks and puts them in a pandas df'''
    url = "http://web.mta.info/developers/data/nyct/turnstile/turnstile_{}.txt"
    dfs = []
    for week_num in week_nums:
        file_url = url.format(week_num)
        dfs.append(pd.read_csv(file_url))
    return pd.concat(dfs)

In [6]:
data = read_data(path)

In [7]:
data.head()

C/A  UNIT       SCP STATION LINENAME DIVISION        DATE      TIME  \
0  A002  R051  02-00-00   59 ST   NQR456      BMT  05/28/2016  00:00:00   
1  A002  R051  02-00-00   59 ST   NQR456      BMT  05/28/2016  04:00:00   
2  A002  R051  02-00-00   59 ST   NQR456      BMT  05/28/2016  08:00:00   
3  A002  R051  02-00-00   59 ST   NQR456      BMT  05/28/2016  12:00:00   
4  A002  R051  02-00-00   59 ST   NQR456      BMT  05/28/2016  16:00:00   

      DESC  ENTRIES    EXITS  
0  REGULAR  5682962  1924340  
1  REGULAR  5683000  1924347  
2  REGULAR  5683030  1924381  
3  REGULAR  5683175  1924474  
4  REGULAR  5683437  1924533

In [8]:
data.tail()

C/A  UNIT       SCP        STATION LINENAME DIVISION        DATE  \
192673  TRAM2  R469  00-05-01  RIT-ROOSEVELT        R      RIT  09/23/2016   
192674  TRAM2  R469  00-05-01  RIT-ROOSEVELT        R      RIT  09/23/2016   
192675  TRAM2  R469  00-05-01  RIT-ROOSEVELT        R      RIT  09/23/2016   
192676  TRAM2  R469  00-05-01  RIT-ROOSEVELT        R      RIT  09/23/2016   
192677  TRAM2  R469  00-05-01  RIT-ROOSEVELT        R      RIT  09/23/2016   

            TIME     DESC  ENTRIES  EXITS  
192673  05:00:00  REGULAR     5554    271  
192674  09:00:00  REGULAR     5554    271  
192675  13:00:00  REGULAR     5554    271  
192676  17:00:00  REGULAR     5554    271  
192677  21:00:00  REGULAR     5554    271

In [9]:
data.DATE.value_counts()
# for each date, give me the number of rows.

09/06/2016    28826
06/07/2016    28816
08/01/2016    28611
06/27/2016    28342
07/05/2016    28268
06/14/2016    28263
07/11/2016    28218
06/13/2016    28200
06/20/2016    28189
07/21/2016    28170
09/16/2016    28160
09/21/2016    28151
09/08/2016    28137
09/01/2016    28116
06/23/2016    28104
07/19/2016    28095
09/12/2016    28074
08/09/2016    28044
06/22/2016    28026
06/29/2016    28022
05/31/2016    28017
07/07/2016    28001
06/08/2016    27986
09/07/2016    27977
09/09/2016    27956
08/04/2016    27917
08/11/2016    27885
07/12/2016    27872
06/01/2016    27863
07/13/2016    27860
              ...  
08/28/2016    27300
08/15/2016    27299
08/16/2016    27296
08/27/2016    27290
07/28/2016    27289
09/15/2016    27283
07/17/2016    27282
07/10/2016    27275
07/08/2016    27274
08/20/2016    27270
07/16/2016    27262
08/13/2016    27262
06/28/2016    27259
08/26/2016    27258
08/07/2016    27258
07/09/2016    27257
07/30/2016    27253
08/03/2016    27249
09/23/2016    27242


Add a column that contains both date and time in datetime format
    


In [10]:
data['DATE_TIME'] = pd.to_datetime(data['DATE'] +' ' + data['TIME'], format = '%m/%d/%Y %H:%M:%S')

In [11]:
data.head()

C/A  UNIT       SCP STATION LINENAME DIVISION        DATE      TIME  \
0  A002  R051  02-00-00   59 ST   NQR456      BMT  05/28/2016  00:00:00   
1  A002  R051  02-00-00   59 ST   NQR456      BMT  05/28/2016  04:00:00   
2  A002  R051  02-00-00   59 ST   NQR456      BMT  05/28/2016  08:00:00   
3  A002  R051  02-00-00   59 ST   NQR456      BMT  05/28/2016  12:00:00   
4  A002  R051  02-00-00   59 ST   NQR456      BMT  05/28/2016  16:00:00   

      DESC  ENTRIES    EXITS           DATE_TIME  
0  REGULAR  5682962  1924340 2016-05-28 00:00:00  
1  REGULAR  5683000  1924347 2016-05-28 04:00:00  
2  REGULAR  5683030  1924381 2016-05-28 08:00:00  
3  REGULAR  5683175  1924474 2016-05-28 12:00:00  
4  REGULAR  5683437  1924533 2016-05-28 16:00:00

Next thing to do is to check the cleanliness of the data:
* There should be one single row for each turnstile (identified by the conjunction of the first four columns) per datetime
* The number of entries from a posterior point in time should be bigger than the number of entries from a previous point in time, for a given trunstile
* There is something about one station actually being two stations, but I will not figure that out.

In [12]:
turnstile = ['C/A', 'UNIT', 'SCP', 'STATION']

In [13]:

data.groupby(turnstile + ['DATE_TIME']).ENTRIES.count().sort_values(ascending = False)

C/A    UNIT  SCP       STATION          DATE_TIME          
N418   R269  01-05-00  BEDFORD-NOSTRAN  2016-08-08 12:00:00    2
R634   R069  00-00-00  NEW LOTS AV      2016-08-20 12:00:00    2
H032   R295  00-05-00  WILSON AV        2016-05-31 20:00:00    2
H009   R235  00-06-01  BEDFORD AV       2016-07-04 16:00:00    2
R634   R069  00-00-00  NEW LOTS AV      2016-08-20 20:00:00    2
B020   R263  00-06-00  AVENUE H         2016-05-30 12:00:00    2
H032   R295  00-05-00  WILSON AV        2016-06-01 00:00:00    2
R634   R069  00-00-00  NEW LOTS AV      2016-08-20 16:00:00    2
N418   R269  01-05-00  BEDFORD-NOSTRAN  2016-08-11 08:00:00    2
                                        2016-08-10 08:00:00    2
H009   R235  00-06-01  BEDFORD AV       2016-07-04 12:00:00    2
N418   R269  01-05-00  BEDFORD-NOSTRAN  2016-08-11 00:00:00    2
H032   R295  00-05-00  WILSON AV        2016-06-01 08:00:00    2
N418   R269  01-05-00  BEDFORD-NOSTRAN  2016-08-10 20:00:00    2
                              

There are several duplicates in the dataset. We need to inspect them to understand what's going on. Once we undesrtand, we can decide what to do with them.


In [14]:
filtre = ((data["C/A"] == "N418") & 
(data["UNIT"] == "R269") & 
(data["SCP"] == "01-05-00") & 
(data["STATION"] == "BEDFORD-NOSTRAN") &
(data["DATE_TIME"].dt.date == datetime(2016, 8, 8).date()))


In [15]:
data[filtre].head(20)

C/A  UNIT       SCP          STATION LINENAME DIVISION        DATE  \
90822  N418  R269  01-05-00  BEDFORD-NOSTRAN        G      IND  08/08/2016   
90823  N418  R269  01-05-00  BEDFORD-NOSTRAN        G      IND  08/08/2016   
90824  N418  R269  01-05-00  BEDFORD-NOSTRAN        G      IND  08/08/2016   
90825  N418  R269  01-05-00  BEDFORD-NOSTRAN        G      IND  08/08/2016   
90826  N418  R269  01-05-00  BEDFORD-NOSTRAN        G      IND  08/08/2016   
90827  N418  R269  01-05-00  BEDFORD-NOSTRAN        G      IND  08/08/2016   
90828  N418  R269  01-05-00  BEDFORD-NOSTRAN        G      IND  08/08/2016   
90829  N418  R269  01-05-00  BEDFORD-NOSTRAN        G      IND  08/08/2016   
90830  N418  R269  01-05-00  BEDFORD-NOSTRAN        G      IND  08/08/2016   
90831  N418  R269  01-05-00  BEDFORD-NOSTRAN        G      IND  08/08/2016   
90832  N418  R269  01-05-00  BEDFORD-NOSTRAN        G      IND  08/08/2016   

           TIME        DESC   ENTRIES     EXITS           DATE_TIME  
90822  00:00:00     REGULAR  83886190         0 2016-08-08 00:00:00  
90823  00:00:00  RECOVR AUD  67108866  33554496 2016-08-08 00:00:00  
90824  04:00:00     REGULAR  83886190         0 2016-08-08 04:00:00  
90825  08:00:00     REGULAR  83886190         0 2016-08-08 08:00:00  
90826  08:00:00  RECOVR AUD  67108866  33554496 2016-08-08 08:00:00  
90827  12:00:00     REGULAR  83886190         0 2016-08-08 12:00:00  
90828  12:00:00  RECOVR AUD  67108866  33554496 2016-08-08 12:00:00  
90829  16:00:00     REGULAR  83886190         0 2016-08-08 16:00:00  
90830  16:00:00  RECOVR AUD  67108866  33554496 2016-08-08 16:00:00  
90831  20:00:00     REGULAR  83886190         0 2016-08-08 20:00:00  
90832  20:00:00  RECOVR AUD  67108866  33554496 2016-08-08 20:00:00

According to the documentation:

DESc     = Represent the "REGULAR" scheduled audit event (Normally occurs every 4 hours)
           1. Audits may occur more that 4 hours due to planning, or troubleshooting activities. 
           2. Additionally, there may be a "RECOVR AUD" entry: This refers to a missed audit that was recovered. 
           
All rows with DESC == 'RECOVR AUD' may be deleted

In [16]:
data_no_dupl = data[data.DESC != 'RECOVR AUD']

In [17]:
data_no_dupl.head()

C/A  UNIT       SCP STATION LINENAME DIVISION        DATE      TIME  \
0  A002  R051  02-00-00   59 ST   NQR456      BMT  05/28/2016  00:00:00   
1  A002  R051  02-00-00   59 ST   NQR456      BMT  05/28/2016  04:00:00   
2  A002  R051  02-00-00   59 ST   NQR456      BMT  05/28/2016  08:00:00   
3  A002  R051  02-00-00   59 ST   NQR456      BMT  05/28/2016  12:00:00   
4  A002  R051  02-00-00   59 ST   NQR456      BMT  05/28/2016  16:00:00   

      DESC  ENTRIES    EXITS           DATE_TIME  
0  REGULAR  5682962  1924340 2016-05-28 00:00:00  
1  REGULAR  5683000  1924347 2016-05-28 04:00:00  
2  REGULAR  5683030  1924381 2016-05-28 08:00:00  
3  REGULAR  5683175  1924474 2016-05-28 12:00:00  
4  REGULAR  5683437  1924533 2016-05-28 16:00:00

Let's see if the duplicates are gone...

In [18]:
data_no_dupl.groupby(turnstile + ['DATE_TIME']).ENTRIES.count().sort_values(ascending = False)

C/A    UNIT  SCP       STATION        DATE_TIME          
TRAM2  R469  00-05-01  RIT-ROOSEVELT  2016-09-23 21:00:00    1
N128   R200  00-05-00  EUCLID AV      2016-08-14 20:00:00    1
                                      2016-08-14 12:00:00    1
                                      2016-08-14 08:00:00    1
                                      2016-08-14 04:00:00    1
                                      2016-08-14 00:00:00    1
                                      2016-08-13 20:00:00    1
                                      2016-08-13 16:00:00    1
                                      2016-08-13 12:00:00    1
                                      2016-08-13 08:00:00    1
                                      2016-08-13 04:00:00    1
                                      2016-08-13 00:00:00    1
                                      2016-08-12 20:00:00    1
                                      2016-08-12 16:00:00    1
                                      2016-08-12 12:00:00   

In [19]:
# drop DESC columns, there's no useful info in there anymore
data_no_dupl = data_no_dupl.drop(['DESC'], axis = 1)

Next: check whether the number of entries (and exits as well if we want to track all transit through a station) actually increases with time.

In [20]:
data_no_dupl.head()

C/A  UNIT       SCP STATION LINENAME DIVISION        DATE      TIME  \
0  A002  R051  02-00-00   59 ST   NQR456      BMT  05/28/2016  00:00:00   
1  A002  R051  02-00-00   59 ST   NQR456      BMT  05/28/2016  04:00:00   
2  A002  R051  02-00-00   59 ST   NQR456      BMT  05/28/2016  08:00:00   
3  A002  R051  02-00-00   59 ST   NQR456      BMT  05/28/2016  12:00:00   
4  A002  R051  02-00-00   59 ST   NQR456      BMT  05/28/2016  16:00:00   

   ENTRIES    EXITS           DATE_TIME  
0  5682962  1924340 2016-05-28 00:00:00  
1  5683000  1924347 2016-05-28 04:00:00  
2  5683030  1924381 2016-05-28 08:00:00  
3  5683175  1924474 2016-05-28 12:00:00  
4  5683437  1924533 2016-05-28 16:00:00

Make a df with one row per turnstile per day. We only keep the first ocurrence of 'entries' for each day, because
What happens during the day is not relevant for a daily analysis. It will be relevant when we want to know peak hours.

In [21]:
data_daily_entries = data_no_dupl.groupby(turnstile+['DATE']).ENTRIES.first().reset_index()

In [22]:
data_daily_entries.head()

C/A  UNIT       SCP STATION        DATE  ENTRIES
0  A002  R051  02-00-00   59 ST  05/28/2016  5682962
1  A002  R051  02-00-00   59 ST  05/29/2016  5683957
2  A002  R051  02-00-00   59 ST  05/30/2016  5684749
3  A002  R051  02-00-00   59 ST  05/31/2016  5685506
4  A002  R051  02-00-00   59 ST  06/01/2016  5687282

In [23]:
data_daily_exits = data_no_dupl.groupby(turnstile+['DATE']).EXITS.first().reset_index()

In [24]:
data_daily_exits.head()

C/A  UNIT       SCP STATION        DATE    EXITS
0  A002  R051  02-00-00   59 ST  05/28/2016  1924340
1  A002  R051  02-00-00   59 ST  05/29/2016  1924599
2  A002  R051  02-00-00   59 ST  05/30/2016  1924791
3  A002  R051  02-00-00   59 ST  05/31/2016  1925050
4  A002  R051  02-00-00   59 ST  06/01/2016  1925601

In [25]:
data_daily_entries[["PREV_DATE", "PREV_ENTRIES"]] = (data_daily_entries
                                                       .groupby(["C/A", "UNIT", "SCP", "STATION"])["DATE", "ENTRIES"]
                                                       .transform(lambda grp: grp.shift(1)))

In [26]:
data_daily_entries.head()

C/A  UNIT       SCP STATION        DATE  ENTRIES   PREV_DATE  PREV_ENTRIES
0  A002  R051  02-00-00   59 ST  05/28/2016  5682962         NaN           NaN
1  A002  R051  02-00-00   59 ST  05/29/2016  5683957  05/28/2016     5682962.0
2  A002  R051  02-00-00   59 ST  05/30/2016  5684749  05/29/2016     5683957.0
3  A002  R051  02-00-00   59 ST  05/31/2016  5685506  05/30/2016     5684749.0
4  A002  R051  02-00-00   59 ST  06/01/2016  5687282  05/31/2016     5685506.0

In [27]:
data_daily_exits["PREV_EXITS"] = data_daily_exits.groupby(turnstile)['EXITS'].transform(lambda grp: grp.shift(1))

In [28]:
data_daily_exits.head()

C/A  UNIT       SCP STATION        DATE    EXITS  PREV_EXITS
0  A002  R051  02-00-00   59 ST  05/28/2016  1924340         NaN
1  A002  R051  02-00-00   59 ST  05/29/2016  1924599   1924340.0
2  A002  R051  02-00-00   59 ST  05/30/2016  1924791   1924599.0
3  A002  R051  02-00-00   59 ST  05/31/2016  1925050   1924791.0
4  A002  R051  02-00-00   59 ST  06/01/2016  1925601   1925050.0

In [29]:
data_daily_entries[['EXITS', 'PREV_EXITS']] = data_daily_exits[['EXITS', 'PREV_EXITS']]

In [30]:
data_daily_entries.head()

C/A  UNIT       SCP STATION        DATE  ENTRIES   PREV_DATE  \
0  A002  R051  02-00-00   59 ST  05/28/2016  5682962         NaN   
1  A002  R051  02-00-00   59 ST  05/29/2016  5683957  05/28/2016   
2  A002  R051  02-00-00   59 ST  05/30/2016  5684749  05/29/2016   
3  A002  R051  02-00-00   59 ST  05/31/2016  5685506  05/30/2016   
4  A002  R051  02-00-00   59 ST  06/01/2016  5687282  05/31/2016   

   PREV_ENTRIES    EXITS  PREV_EXITS  
0           NaN  1924340         NaN  
1     5682962.0  1924599   1924340.0  
2     5683957.0  1924791   1924599.0  
3     5684749.0  1925050   1924791.0  
4     5685506.0  1925601   1925050.0

'data_daily_entries' now is a df which permits unique identification of each turnstile (first 4 columns), the date, the previous date, the entries and exits and previous entries and exits. 
Next: drop rows with NaN

In [31]:
data_daily_entries.dropna(subset=["PREV_DATE"], axis=0, inplace=True)

In [32]:
data_daily_entries.head()

C/A  UNIT       SCP STATION        DATE  ENTRIES   PREV_DATE  \
1  A002  R051  02-00-00   59 ST  05/29/2016  5683957  05/28/2016   
2  A002  R051  02-00-00   59 ST  05/30/2016  5684749  05/29/2016   
3  A002  R051  02-00-00   59 ST  05/31/2016  5685506  05/30/2016   
4  A002  R051  02-00-00   59 ST  06/01/2016  5687282  05/31/2016   
5  A002  R051  02-00-00   59 ST  06/02/2016  5689073  06/01/2016   

   PREV_ENTRIES    EXITS  PREV_EXITS  
1     5682962.0  1924599   1924340.0  
2     5683957.0  1924791   1924599.0  
3     5684749.0  1925050   1924791.0  
4     5685506.0  1925601   1925050.0  
5     5687282.0  1926208   1925601.0

In [33]:
crazy_turnstiles = data_daily_entries[data_daily_entries["ENTRIES"] < data_daily_entries["PREV_ENTRIES"]]

In [34]:
crazy_turnstiles.head()

C/A  UNIT       SCP     STATION        DATE    ENTRIES   PREV_DATE  \
3725  A010  R080  00-00-05  57 ST-7 AV  07/15/2016       1924  07/14/2016   
4035  A011  R080  01-00-00  57 ST-7 AV  05/29/2016  887922841  05/28/2016   
4036  A011  R080  01-00-00  57 ST-7 AV  05/30/2016  887921264  05/29/2016   
4037  A011  R080  01-00-00  57 ST-7 AV  05/31/2016  887919921  05/30/2016   
4038  A011  R080  01-00-00  57 ST-7 AV  06/01/2016  887916819  05/31/2016   

      PREV_ENTRIES      EXITS   PREV_EXITS  
3725    10780997.0         59    3955742.0  
4035   887924524.0  493656275  493658400.0  
4036   887922841.0  493654483  493656275.0  
4037   887921264.0  493652865  493654483.0  
4038   887919921.0  493649969  493652865.0

In [35]:
# crazy turnstile days for enterung
len(crazy_turnstiles)

4298

In [36]:
crazy_turnstiles_exit = data_daily_entries[data_daily_entries["EXITS"] < data_daily_entries["PREV_EXITS"]]
crazy_turnstiles_exit.head()

C/A  UNIT       SCP     STATION        DATE    ENTRIES   PREV_DATE  \
3725  A010  R080  00-00-05  57 ST-7 AV  07/15/2016       1924  07/14/2016   
4035  A011  R080  01-00-00  57 ST-7 AV  05/29/2016  887922841  05/28/2016   
4036  A011  R080  01-00-00  57 ST-7 AV  05/30/2016  887921264  05/29/2016   
4037  A011  R080  01-00-00  57 ST-7 AV  05/31/2016  887919921  05/30/2016   
4038  A011  R080  01-00-00  57 ST-7 AV  06/01/2016  887916819  05/31/2016   

      PREV_ENTRIES      EXITS   PREV_EXITS  
3725    10780997.0         59    3955742.0  
4035   887924524.0  493656275  493658400.0  
4036   887922841.0  493654483  493656275.0  
4037   887921264.0  493652865  493654483.0  
4038   887919921.0  493649969  493652865.0

In [37]:
# crazy turnstile days exit
len(crazy_turnstiles_exit)

3676

In [38]:
# total number of turnstile days
len(data_daily_entries)

538364

In [39]:
# number of turnstiles
len(data_daily_entries.groupby(turnstile))

4626

In [40]:
# check if the crazy turnstile for exiting and entering coincide
ct_entries = crazy_turnstiles.groupby(turnstile).groups.keys()
len(ct_entries)

225

In [41]:
ct_exits = crazy_turnstiles_exit.groupby(turnstile).groups.keys()
len(ct_exits)

226

In [42]:
always_crazy = set.intersection(set(ct_entries), set(ct_exits))

In [43]:
len(always_crazy)

200

Decision to drop crazy turnstiles or not:
* We have 225 turnstiles that can go crazy at the entries, we have 226 turnstiles that can go crazy at the exit. 200 turnstiles are in both sets.
* total number of turnstiles is 4626
* the crazy turnstiles make up approximately 5% of the turnstiles (crazy = 251, total = 4626, percentage = 5%)

_We are going the drop the crazy trunstiles_

In [44]:
# identify crazy turnstiles:
all_crazies = list(set(list(ct_entries) + list(ct_exits)))

In [45]:
all_crazies

[('J030', 'R005', '00-00-02', '85 ST-FOREST PK'),
 ('A011', 'R080', '01-00-00', '57 ST-7 AV'),
 ('E001', 'R368', '00-00-03', '9 AV'),
 ('R501', 'R054', '00-00-04', '5 AVE'),
 ('PTH16', 'R550', '01-01-03', 'LACKAWANNA'),
 ('R148', 'R033', '01-00-01', 'TIMES SQ-42 ST'),
 ('PTH18', 'R549', '01-00-08', 'NEWARK BM BW'),
 ('PTH16', 'R550', '01-01-04', 'LACKAWANNA'),
 ('A025', 'R023', '01-03-02', '34 ST-HERALD SQ'),
 ('N001', 'R173', '01-06-03', 'INWOOD-207 ST'),
 ('R194', 'R040', '00-06-01', '231 ST'),
 ('N039', 'R251', '01-00-01', '96 ST'),
 ('R527', 'R122', '00-05-01', '90 ST-ELMHURST'),
 ('N401', 'R360', '00-00-01', '21 ST'),
 ('R246', 'R177', '00-00-02', '68ST-HUNTER CO'),
 ('N089', 'R139', '00-03-00', 'CANAL ST'),
 ('N506', 'R022', '00-03-01', '34 ST-HERALD SQ'),
 ('R161A', 'R452', '01-06-02', '72 ST'),
 ('R524', 'R347', '00-05-00', '69 ST'),
 ('R127', 'R105', '00-00-00', '14 ST'),
 ('R169', 'R168', '01-07-00', '96 ST'),
 ('N016A', 'R296', '00-00-01', '163 ST-AMSTERDM'),
 ('A011', 'R080

In [46]:
stations_with_ct = []
for ts in all_crazies:
    stations_with_ct.append(ts[3])

In [47]:
len(set(stations_with_ct))

142

There are 142 stations where at a certain point there was a crazy turnstile, out of a total of 376 stations.

In [48]:
# total number of stations:
len(data.STATION.unique())

376

We decided to drop the crazy turnstiles altogether, since at first glance they are well spread among the stations. 
Future work might include determining distribution of crazy turnstiles in trainstations, to better understand the impact of dropping these data. There might be higher concentration of crazy turnstiles in certain stations. For this to be possible, it is necessary to determine the number of turnstiles in each station, determine how many of these are faulty, and make sure that dropping data from crazy turnstiles has a uniform impast on the data from each station.

In [49]:
data_daily = data_daily_entries
data_daily

C/A  UNIT       SCP        STATION        DATE  ENTRIES   PREV_DATE  \
1        A002  R051  02-00-00          59 ST  05/29/2016  5683957  05/28/2016   
2        A002  R051  02-00-00          59 ST  05/30/2016  5684749  05/29/2016   
3        A002  R051  02-00-00          59 ST  05/31/2016  5685506  05/30/2016   
4        A002  R051  02-00-00          59 ST  06/01/2016  5687282  05/31/2016   
5        A002  R051  02-00-00          59 ST  06/02/2016  5689073  06/01/2016   
6        A002  R051  02-00-00          59 ST  06/03/2016  5690839  06/02/2016   
7        A002  R051  02-00-00          59 ST  06/04/2016  5692587  06/03/2016   
8        A002  R051  02-00-00          59 ST  06/05/2016  5693604  06/04/2016   
9        A002  R051  02-00-00          59 ST  06/06/2016  5694361  06/05/2016   
10       A002  R051  02-00-00          59 ST  06/07/2016  5696129  06/06/2016   
11       A002  R051  02-00-00          59 ST  06/08/2016  5697918  06/07/2016   
12       A002  R051  02-00-00          59 ST  06/09/2016  5699775  06/08/2016   
13       A002  R051  02-00-00          59 ST  06/10/2016  5701611  06/09/2016   
14       A002  R051  02-00-00          59 ST  06/11/2016  5703420  06/10/2016   
15       A002  R051  02-00-00          59 ST  06/12/2016  5704439  06/11/2016   
16       A002  R051  02-00-00          59 ST  06/13/2016  5704958  06/12/2016   
17       A002  R051  02-00-00          59 ST  06/14/2016  5704958  06/13/2016   
18       A002  R051  02-00-00          59 ST  06/15/2016  5704958  06/14/2016   
19       A002  R051  02-00-00          59 ST  06/16/2016  5704958  06/15/2016   
20       A002  R051  02-00-00          59 ST  06/18/2016  5704958  06/16/2016   
21       A002  R051  02-00-00          59 ST  06/19/2016  5704958  06/18/2016   
22       A002  R051  02-00-00          59 ST  06/20/2016  5705337  06/19/2016   
23       A002  R051  02-00-00          59 ST  06/21/2016  5706844  06/20/2016   
24       A002  R051  02-00-00          59 ST  06/22/2016  5708543  06/21/2016   
25       A002  R051  02-00-00          59 ST  06/23/2016  5710209  06/22/2016   
26       A002  R051  02-00-00          59 ST  06/24/2016  5711852  06/23/2016   
27       A002  R051  02-00-00          59 ST  06/25/2016  5713268  06/24/2016   
28       A002  R051  02-00-00          59 ST  06/26/2016  5714130  06/25/2016   
29       A002  R051  02-00-00          59 ST  06/27/2016  5714706  06/26/2016   
30       A002  R051  02-00-00          59 ST  06/28/2016  5716372  06/27/2016   
...       ...   ...       ...            ...         ...      ...         ...   
542963  TRAM2  R469  00-05-01  RIT-ROOSEVELT  08/25/2016     5554  08/24/2016   
542964  TRAM2  R469  00-05-01  RIT-ROOSEVELT  08/26/2016     5554  08/25/2016   
542965  TRAM2  R469  00-05-01  RIT-ROOSEVELT  08/27/2016     5554  08/26/2016   
542966  TRAM2  R469  00-05-01  RIT-ROOSEVELT  08/28/2016     5554  08/27/2016   
542967  TRAM2  R469  00-05-01  RIT-ROOSEVELT  08/29/2016     5554  08/28/2016   
542968  TRAM2  R469  00-05-01  RIT-ROOSEVELT  08/30/2016     5554  08/29/2016   
542969  TRAM2  R469  00-05-01  RIT-ROOSEVELT  08/31/2016     5554  08/30/2016   
542970  TRAM2  R469  00-05-01  RIT-ROOSEVELT  09/01/2016     5554  08/31/2016   
542971  TRAM2  R469  00-05-01  RIT-ROOSEVELT  09/02/2016     5554  09/01/2016   
542972  TRAM2  R469  00-05-01  RIT-ROOSEVELT  09/03/2016     5554  09/02/2016   
542973  TRAM2  R469  00-05-01  RIT-ROOSEVELT  09/04/2016     5554  09/03/2016   
542974  TRAM2  R469  00-05-01  RIT-ROOSEVELT  09/05/2016     5554  09/04/2016   
542975  TRAM2  R469  00-05-01  RIT-ROOSEVELT  09/06/2016     5554  09/05/2016   
542976  TRAM2  R469  00-05-01  RIT-ROOSEVELT  09/07/2016     5554  09/06/2016   
542977  TRAM2  R469  00-05-01  RIT-ROOSEVELT  09/08/2016     5554  09/07/2016   
542978  TRAM2  R469  00-05-01  RIT-ROOSEVELT  09/09/2016     5554  09/08/2016   
542979  TRAM2  R469  00-05-01  RIT-ROOSEVELT  09/10/2016     5554  09/09/2016   
542980  TRAM2  R469  00-05-01  RIT-ROOSEV

In [50]:
# drop the crazy turnstiles:
for ts in all_crazies:
    mask = ((data_daily["C/A"] == ts[0]) & (data_daily["UNIT"] == ts[1]) & 
            (data_daily["SCP"] == ts[2]) & (data_daily["STATION"] == ts[3]))
    data_daily.drop(data_daily[mask].index, inplace=True)

In [51]:
data_daily[(data_daily['C/A']=='A002') & (data_daily.SCP == '02-00-00')]

C/A  UNIT       SCP STATION        DATE  ENTRIES   PREV_DATE  \
1    A002  R051  02-00-00   59 ST  05/29/2016  5683957  05/28/2016   
2    A002  R051  02-00-00   59 ST  05/30/2016  5684749  05/29/2016   
3    A002  R051  02-00-00   59 ST  05/31/2016  5685506  05/30/2016   
4    A002  R051  02-00-00   59 ST  06/01/2016  5687282  05/31/2016   
5    A002  R051  02-00-00   59 ST  06/02/2016  5689073  06/01/2016   
6    A002  R051  02-00-00   59 ST  06/03/2016  5690839  06/02/2016   
7    A002  R051  02-00-00   59 ST  06/04/2016  5692587  06/03/2016   
8    A002  R051  02-00-00   59 ST  06/05/2016  5693604  06/04/2016   
9    A002  R051  02-00-00   59 ST  06/06/2016  5694361  06/05/2016   
10   A002  R051  02-00-00   59 ST  06/07/2016  5696129  06/06/2016   
11   A002  R051  02-00-00   59 ST  06/08/2016  5697918  06/07/2016   
12   A002  R051  02-00-00   59 ST  06/09/2016  5699775  06/08/2016   
13   A002  R051  02-00-00   59 ST  06/10/2016  5701611  06/09/2016   
14   A002  R051  02-00-00   59 ST  06/11/2016  5703420  06/10/2016   
15   A002  R051  02-00-00   59 ST  06/12/2016  5704439  06/11/2016   
16   A002  R051  02-00-00   59 ST  06/13/2016  5704958  06/12/2016   
17   A002  R051  02-00-00   59 ST  06/14/2016  5704958  06/13/2016   
18   A002  R051  02-00-00   59 ST  06/15/2016  5704958  06/14/2016   
19   A002  R051  02-00-00   59 ST  06/16/2016  5704958  06/15/2016   
20   A002  R051  02-00-00   59 ST  06/18/2016  5704958  06/16/2016   
21   A002  R051  02-00-00   59 ST  06/19/2016  5704958  06/18/2016   
22   A002  R051  02-00-00   59 ST  06/20/2016  5705337  06/19/2016   
23   A002  R051  02-00-00   59 ST  06/21/2016  5706844  06/20/2016   
24   A002  R051  02-00-00   59 ST  06/22/2016  5708543  06/21/2016   
25   A002  R051  02-00-00   59 ST  06/23/2016  5710209  06/22/2016   
26   A002  R051  02-00-00   59 ST  06/24/2016  5711852  06/23/2016   
27   A002  R051  02-00-00   59 ST  06/25/2016  5713268  06/24/2016   
28   A002  R051  02-00-00   59 ST  06/26/2016  5714130  06/25/2016   
29   A002  R051  02-00-00   59 ST  06/27/2016  5714706  06/26/2016   
30   A002  R051  02-00-00   59 ST  06/28/2016  5716372  06/27/2016   
..    ...   ...       ...     ...         ...      ...         ...   
88   A002  R051  02-00-00   59 ST  08/25/2016  5796665  08/24/2016   
89   A002  R051  02-00-00   59 ST  08/26/2016  5797976  08/25/2016   
90   A002  R051  02-00-00   59 ST  08/27/2016  5799442  08/26/2016   
91   A002  R051  02-00-00   59 ST  08/28/2016  5800252  08/27/2016   
92   A002  R051  02-00-00   59 ST  08/29/2016  5800934  08/28/2016   
93   A002  R051  02-00-00   59 ST  08/30/2016  5802514  08/29/2016   
94   A002  R051  02-00-00   59 ST  08/31/2016  5804071  08/30/2016   
95   A002  R051  02-00-00   59 ST  09/01/2016  5805638  08/31/2016   
96   A002  R051  02-00-00   59 ST  09/02/2016  5807232  09/01/2016   
97   A002  R051  02-00-00   59 ST  09/03/2016  5808846  09/02/2016   
98   A002  R051  02-00-00   59 ST  09/04/2016  5809667  09/03/2016   
99   A002  R051  02-00-00   59 ST  09/05/2016  5810301  09/04/2016   
100  A002  R051  02-00-00   59 ST  09/06/2016  5810983  09/05/2016   
101  A002  R051  02-00-00   59 ST  09/07/2016  5812369  09/06/2016   
102  A002  R051  02-00-00   59 ST  09/08/2016  5814038  09/07/2016   
103  A002  R051  02-00-00   59 ST  09/09/2016  5815801  09/08/2016   
104  A002  R051  02-00-00   59 ST  09/10/2016  5817505  09/09/2016   
105  A002  R051  02-00-00   59 ST  09/11/2016  5818403  09/10/2016   
106  A002  R051  02-00-00   59 ST  09/12/2016  5819116  09/11/2016   
107  A002  R051  02-00-00   59 ST  09/13/2016  5820776  09/12/2016   
108  A002  R051  02-00-00   59 ST  09/14/2016  5822527  09/13/2016   
109  A002  R051  02-00-00   59 ST  09/15/2016  5824271  09/14/2016   
110  A002  R051  02-00-00   59 ST  09/16/2016  5826092  09/15/2016   
111  A002  R051  02-00-00   59 ST  09/17/2016  5827822  09/16/2016   
112  A002  R051  02-00-00   59 ST  09/18/2016  5828891  09/17/2016   
113  A002  R051 

In [52]:
data_daily['SALDO_ENTRIES'] = data_daily['ENTRIES']-data_daily['PREV_ENTRIES']
data_daily['SALDO_EXITS'] = data_daily['EXITS']-data_daily['PREV_EXITS']
data_daily['TRANSITING'] = data_daily['SALDO_ENTRIES'] + data_daily['SALDO_EXITS'] 

In [53]:
data_daily.head()

C/A  UNIT       SCP STATION        DATE  ENTRIES   PREV_DATE  \
1  A002  R051  02-00-00   59 ST  05/29/2016  5683957  05/28/2016   
2  A002  R051  02-00-00   59 ST  05/30/2016  5684749  05/29/2016   
3  A002  R051  02-00-00   59 ST  05/31/2016  5685506  05/30/2016   
4  A002  R051  02-00-00   59 ST  06/01/2016  5687282  05/31/2016   
5  A002  R051  02-00-00   59 ST  06/02/2016  5689073  06/01/2016   

   PREV_ENTRIES    EXITS  PREV_EXITS  SALDO_ENTRIES  SALDO_EXITS  TRANSITING  
1     5682962.0  1924599   1924340.0          995.0        259.0      1254.0  
2     5683957.0  1924791   1924599.0          792.0        192.0       984.0  
3     5684749.0  1925050   1924791.0          757.0        259.0      1016.0  
4     5685506.0  1925601   1925050.0         1776.0        551.0      2327.0  
5     5687282.0  1926208   1925601.0         1791.0        607.0      2398.0

data_daily is a df that has column 'TRANSITING' which contains the number of people going through a turnstile on a certain day

We still get some abnormal datapoints. We will just cut off what doesn't make sense.

In [54]:
data_daily.SALDO_ENTRIES.sort_values(ascending=False).head(100)

165688    1.905098e+09
205854    3.342257e+08
76276     1.509941e+08
456164    1.174400e+08
397483    1.174397e+08
48995     1.006632e+08
197074    6.710886e+07
480087    6.697559e+07
55185     6.633143e+07
288295    6.598917e+07
8717      6.547201e+07
85290     6.047959e+07
157946    4.920656e+07
412660    1.677688e+07
142480    2.402833e+06
161062    1.293359e+06
472342    3.931740e+05
196955    2.618530e+05
177356    2.615710e+05
69174     5.407000e+04
278101    1.629700e+04
380408    1.413200e+04
157987    1.328600e+04
161527    1.314400e+04
115050    1.251800e+04
380523    1.184000e+04
372975    1.108900e+04
380293    1.053600e+04
277189    1.043200e+04
286086    1.006000e+04
              ...     
416516    8.816000e+03
416944    8.773000e+03
419679    8.768000e+03
416943    8.762000e+03
416951    8.727000e+03
416558    8.707000e+03
416556    8.705000e+03
419631    8.687000e+03
416508    8.676000e+03
416950    8.649000e+03
419678    8.632000e+03
416947    8.630000e+03
419574    8

In [55]:
data_daily = data_daily[data_daily['SALDO_ENTRIES'] < 10000]

In [56]:
data_daily.to_csv('/Users/aleksandra/ds/metis/Project-Benson/Data/data_daily.csv')

Next: make a df called 'stations_transit' that has a column of stations and a column of dates and a column of people transiting.

In [57]:
list_of_stations =list(data_daily.STATION.unique())

In [58]:
list_of_stations

['59 ST',
 '5 AV/59 ST',
 '57 ST-7 AV',
 '49 ST',
 'TIMES SQ-42 ST',
 '34 ST-HERALD SQ',
 '28 ST',
 '23 ST',
 '14 ST-UNION SQ',
 '8 ST-NYU',
 'PRINCE ST',
 'CANAL ST',
 'CITY HALL',
 'CORTLANDT ST',
 'RECTOR ST',
 'WHITEHALL S-FRY',
 'DELANCEY/ESSEX',
 'BOWERY',
 'CHAMBERS ST',
 'FULTON ST',
 'BROAD ST',
 '7 AV',
 'PARK PLACE',
 'BOTANIC GARDEN',
 'PROSPECT PARK',
 'PARKSIDE AV',
 'CHURCH AV',
 'BEVERLEY ROAD',
 'CORTELYOU RD',
 'NEWKIRK PLAZA',
 'AVENUE H',
 'AVENUE J',
 'AVENUE M',
 'KINGS HWY',
 'AVENUE U',
 'NECK RD',
 'SHEEPSHEAD BAY',
 'BRIGHTON BEACH',
 'OCEAN PKWY',
 'BOROUGH HALL',
 'JAY ST-METROTEC',
 'DEKALB AV',
 'ATL AV-BARCLAY',
 'UNION ST',
 '4AV-9 ST',
 'PROSPECT AV',
 '25 ST',
 '36 ST',
 '45 ST',
 '53 ST',
 'BAY RIDGE AV',
 '77 ST',
 '86 ST',
 'BAY RIDGE-95 ST',
 '8 AV',
 'FT HAMILTON PKY',
 'NEW UTRECHT AV',
 '18 AV',
 '20 AV',
 'BAY PKWY',
 '9 AV',
 '50 ST',
 '55 ST',
 '71 ST',
 '79 ST',
 '25 AV',
 'BAY 50 ST',
 'CONEY IS-STILLW',
 'W 8 ST-AQUARIUM',
 '6 AV',
 '3 AV'

In [59]:
stations_transit = data_daily.groupby(['STATION', 'DATE'])['TRANSITING'].sum()
# stations_transit = data_daily.groupby(['STATION', 'DATE'], as_index=False)['TRANSITING'].sum()

In [60]:
stations_transit.head()

STATION  DATE      
1 AV     05/29/2016    30828.0
         05/30/2016    28051.0
         05/31/2016    23702.0
         06/01/2016    42811.0
         06/02/2016    43633.0
Name: TRANSITING, dtype: float64

In [61]:
stations_transit = pd.DataFrame(stations_transit)

In [62]:
stations_transit['index'] = stations_transit.index
stations_transit.head()

TRANSITING               index
STATION DATE                                      
1 AV    05/29/2016     30828.0  (1 AV, 05/29/2016)
        05/30/2016     28051.0  (1 AV, 05/30/2016)
        05/31/2016     23702.0  (1 AV, 05/31/2016)
        06/01/2016     42811.0  (1 AV, 06/01/2016)
        06/02/2016     43633.0  (1 AV, 06/02/2016)

In [63]:
stations_transit[['STATION', 'DATE']] = stations_transit['index'].apply(pd.Series)

In [64]:
stations_transit.drop('index', 1, inplace=True)

In [65]:
stations_transit.head()

TRANSITING STATION        DATE
STATION DATE                                      
1 AV    05/29/2016     30828.0    1 AV  05/29/2016
        05/30/2016     28051.0    1 AV  05/30/2016
        05/31/2016     23702.0    1 AV  05/31/2016
        06/01/2016     42811.0    1 AV  06/01/2016
        06/02/2016     43633.0    1 AV  06/02/2016

Make a column containing the day of the week

In [66]:
stations_transit['DATE'] = pd.to_datetime(stations_transit['DATE'])
stations_transit['DAY_OF_WEEK'] = stations_transit['DATE'].dt.weekday_name

In [67]:
stations_transit.head()

TRANSITING STATION       DATE DAY_OF_WEEK
STATION DATE                                                 
1 AV    05/29/2016     30828.0    1 AV 2016-05-29      Sunday
        05/30/2016     28051.0    1 AV 2016-05-30      Monday
        05/31/2016     23702.0    1 AV 2016-05-31     Tuesday
        06/01/2016     42811.0    1 AV 2016-06-01   Wednesday
        06/02/2016     43633.0    1 AV 2016-06-02    Thursday

In [68]:
stations_transit.reset_index(inplace=True, drop=True)

In [69]:
stations_transit.head()

TRANSITING STATION       DATE DAY_OF_WEEK
0     30828.0    1 AV 2016-05-29      Sunday
1     28051.0    1 AV 2016-05-30      Monday
2     23702.0    1 AV 2016-05-31     Tuesday
3     42811.0    1 AV 2016-06-01   Wednesday
4     43633.0    1 AV 2016-06-02    Thursday

In [70]:
stations_transit.to_csv('/Users/aleksandra/ds/metis/Project-Benson/Data/per_station.csv')

Make df that sums transit per station per day of the week.

In [71]:
transit_dw_stations = stations_transit.groupby(['STATION', 'DAY_OF_WEEK'])['TRANSITING'].sum()

In [72]:
transit_dw_stations = pd.DataFrame(transit_dw_stations)

In [73]:
transit_dw_stations.head()

TRANSITING
STATION DAY_OF_WEEK            
1 AV    Friday         692467.0
        Monday         433302.0
        Saturday       675612.0
        Sunday         527998.0
        Thursday       691732.0

In [74]:
transit_dw_stations['index'] = transit_dw_stations.index
transit_dw_stations.head()

TRANSITING             index
STATION DAY_OF_WEEK                              
1 AV    Friday         692467.0    (1 AV, Friday)
        Monday         433302.0    (1 AV, Monday)
        Saturday       675612.0  (1 AV, Saturday)
        Sunday         527998.0    (1 AV, Sunday)
        Thursday       691732.0  (1 AV, Thursday)

In [75]:
transit_dw_stations[['STATION', 'DAY_OF_WEEK']] = transit_dw_stations['index'].apply(pd.Series)

In [76]:
transit_dw_stations.head()

TRANSITING             index STATION DAY_OF_WEEK
STATION DAY_OF_WEEK                                                  
1 AV    Friday         692467.0    (1 AV, Friday)    1 AV      Friday
        Monday         433302.0    (1 AV, Monday)    1 AV      Monday
        Saturday       675612.0  (1 AV, Saturday)    1 AV    Saturday
        Sunday         527998.0    (1 AV, Sunday)    1 AV      Sunday
        Thursday       691732.0  (1 AV, Thursday)    1 AV    Thursday

In [77]:
transit_dw_stations.drop('index', 1, inplace=True)

In [78]:
transit_dw_stations.reset_index(inplace=True, drop=True)

In [79]:
transit_dw_stations.head(7)

TRANSITING STATION DAY_OF_WEEK
0    692467.0    1 AV      Friday
1    433302.0    1 AV      Monday
2    675612.0    1 AV    Saturday
3    527998.0    1 AV      Sunday
4    691732.0    1 AV    Thursday
5    600885.0    1 AV     Tuesday
6    675432.0    1 AV   Wednesday

In [80]:
transit_dw_stations.to_csv('/Users/aleksandra/ds/metis/Project-Benson/Data/station_dow.csv')

Now we have three csv's:
* data_daily with everything
* stations_transit with transit per station per day
* transit_dw_stations with transit per station per weekday

We are going to add a column with the weather on a certain day

In [1]:
len(data_daily)

NameError: name 'data_daily' is not defined

In [81]:
weather_data = pd.read_csv('/Users/aleksandra/ds/metis/Project-Benson/Data/NYC_rainy_days.csv')

In [82]:
weather_data.head()

Unnamed: 0            STATION                STATION_NAME        DATE  PRCP
0        4645  GHCND:USW00094728  NY CITY CENTRAL PARK NY US  2016-06-27  0.45
1        4646  GHCND:USW00094728  NY CITY CENTRAL PARK NY US  2016-06-28  0.12
2        4647  GHCND:USW00094728  NY CITY CENTRAL PARK NY US  2016-06-29  0.01
3        4648  GHCND:USW00094728  NY CITY CENTRAL PARK NY US  2016-06-30  0.00
4        4649  GHCND:USW00094728  NY CITY CENTRAL PARK NY US  2016-07-01  0.83

In [83]:
data_daily.head()

C/A  UNIT       SCP STATION        DATE  ENTRIES   PREV_DATE  \
1  A002  R051  02-00-00   59 ST  05/29/2016  5683957  05/28/2016   
2  A002  R051  02-00-00   59 ST  05/30/2016  5684749  05/29/2016   
3  A002  R051  02-00-00   59 ST  05/31/2016  5685506  05/30/2016   
4  A002  R051  02-00-00   59 ST  06/01/2016  5687282  05/31/2016   
5  A002  R051  02-00-00   59 ST  06/02/2016  5689073  06/01/2016   

   PREV_ENTRIES    EXITS  PREV_EXITS  SALDO_ENTRIES  SALDO_EXITS  TRANSITING  
1     5682962.0  1924599   1924340.0          995.0        259.0      1254.0  
2     5683957.0  1924791   1924599.0          792.0        192.0       984.0  
3     5684749.0  1925050   1924791.0          757.0        259.0      1016.0  
4     5685506.0  1925601   1925050.0         1776.0        551.0      2327.0  
5     5687282.0  1926208   1925601.0         1791.0        607.0      2398.0

We want to add the column PRCP to the data_daily dataframe, mathcing on date. First we need to cnvert the DATE columns in both dataframes to the same format.

In [84]:
data_daily['DATE_2'] = pd.to_datetime(data_daily['DATE'], format = '%m/%d/%Y')

//anaconda/envs/py3/lib/python3.6/site-packages/ipykernel/__main__.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  if __name__ == '__main__':


In [85]:
small_weather = weather_data[['DATE', 'PRCP']]
small_weather.head()

DATE  PRCP
0  2016-06-27  0.45
1  2016-06-28  0.12
2  2016-06-29  0.01
3  2016-06-30  0.00
4  2016-07-01  0.83

In [86]:
data_daily_weather = pd.merge(data_daily, small_weather, left_on = 'DATE_2', right_on='DATE', how = 'left')

In [87]:
data_daily_weather['PRCP'].isnull().sum()

509155

In [88]:
len(data_daily_weather['PRCP'])

509155

25% difference is due to the fact that the mta data we used is for June-September, while the weather data is from July-September. We just dropped June.

In [89]:
data_daily_weather = data_daily.dropna(axis = 0)

In [90]:
data_daily_weather.to_csv('/Users/aleksandra/ds/metis/Project-Benson/Data/data_daily_weather.csv')